codec/src/main/java/io/netty/handler/codec/compression/CompressionEncoder.java

```diff
@@ -0,0 +1,182 @@
+import io.netty.channel.ChannelPipeline;
+import io.netty.channel.ChannelPromise;
+import io.netty.channel.ChannelPromiseNotifier;
+import io.netty.handler.codec.MessageToByteEncoder;
+import io.netty.util.concurrent.EventExecutor;
+import io.netty.util.internal.ObjectUtil;
+import io.netty.util.internal.OneTimeTask;
+
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Compresses a {@link ByteBuf} using one of compression algorithms.
+ */
+public abstract class CompressionEncoder extends MessageToByteEncoder<ByteBuf> {
+    /**
+     * Default delay time for close. See method {@link #closeDelay(int)}.
+     */
+    public static final int DEFAULT_CLOSE_DELAY = 10;
+
+    private int closeDelay;
+
+    private final CompressionFormat format;
+
+    /**
+     * Used to interact with its {@link ChannelPipeline} and other handlers.
+     */
+    private volatile ChannelHandlerContext ctx;
+
+    /**
+     * Default constructor for {@link CompressionEncoder} which will try to use a direct
+     * {@link ByteBuf} as a target for the encoded messages.
+     */
+    protected CompressionEncoder(CompressionFormat format) {
+        this(format, true);
+    }
+
+    /**
+     * Creates an instance with specified preference of a target {@link ByteBuf} for the encoded messages.
+     * It uses if compression algorithm needs a byte array as a target.
+     *
+     * @param preferDirect {@code true} if a direct {@link ByteBuf} should be tried to be used as a target
+     *                     for the encoded messages. If {@code false} is used it will allocate a heap
+     *                     {@link ByteBuf}, which is backed by an byte array.
+     */
+    protected CompressionEncoder(CompressionFormat format, boolean preferDirect) {
+        super(preferDirect);
+        ObjectUtil.checkNotNull(format, "format");
+        this.format = format;
+        this.closeDelay = DEFAULT_CLOSE_DELAY;
+    }
+
+    /**
+     * Returns a format of current compression algorithm.
+     */
+    public final CompressionFormat format() {
+        return format;
+    }
+
+    /**
+     * Set amount of time to be delayed before the ensuring that the channel is closed
+     * after the call of {@link #close(ChannelHandlerContext, ChannelPromise)}.
+     *
+     * The default value is {@link #DEFAULT_CLOSE_DELAY}.
+     *
+     * @param delay time in seconds.
+     */
+    public final void closeDelay(int delay) {
+        closeDelay = delay;
+    }
+
+    /**
+     * Returns {@code true} if and only if the end of the compressed stream has been reached.
+     */
+    public abstract boolean isClosed();
+
+    /**
+     * Close this {@link CompressionEncoder} and so finish the encoding.
+     *
+     * The returned {@link ChannelFuture} will be notified once the operation completes.
+     *
+     * Note: for some {@link CompressionFormat}'s which does not support close operation
+     * this method could return failed {@link ChannelFuture}.
```

Comment:
Should I add javadoc to all overriden methods, that they return failed ChannelFuture? Like this https://github.com/netty/netty/pull/2527/files#diff-3411b3b6db61cc9b3639d77d4e421e7eR196